pkg/controller/container-runtime-config/container_runtime_config_controller.go

```diff
@@ -607,6 +618,94 @@ func (ctrl *Controller) mergeConfigChanges(origFile *igntypes.File, cfg *mcfgv1.
+// resync.
+func (ctrl *Controller) createCapabilitiesMC() error {
+	var configMapName = "crio-capabilities"
+
+	// Check if the crio-capabilities config map exists in the openshift-machine-config-operator namespace
+	capCM, err := ctrl.kubeClient.CoreV1().ConfigMaps(ctrl.namespace).Get(context.TODO(), configMapName, metav1.GetOptions{})
+	capCMIsNotFound := errors.IsNotFound(err)
+	if err != nil && !capCMIsNotFound {
+		return fmt.Errorf("error checking for %s config map: %v", configMapName, err)
+	}
+	// If the crio-capabilities config map exists, that means the capabilities MC was already created
+	// so we should not create this MC again and return
+	if capCM != nil && !capCMIsNotFound {
+		return nil
+	}
+	// Create the config map for capabilities so we know that the capabilities MC has been created
+	capCM.Name = configMapName
+	capCM.Namespace = ctrl.namespace
+	if _, err := ctrl.kubeClient.CoreV1().ConfigMaps(ctrl.namespace).Create(context.TODO(), capCM, metav1.CreateOptions{}); err != nil {
+		return fmt.Errorf("error creating %s config map: %v", configMapName, err)
+	}
+
+	sel, err := metav1.LabelSelectorAsSelector(metav1.AddLabelToSelector(&metav1.LabelSelector{}, builtInLabelKey, ""))
+	if err != nil {
+		return err
+	}
+	// Find all the MachineConfigPools
+	mcpPoolsAll, err := ctrl.mcpLister.List(sel)
+	if err != nil {
+		return err
+	}
+
+	// Create the capabilities MC for all the available pools
+	for _, pool := range mcpPoolsAll {
+		managedKey := getManagedKeyCaps(pool)
+		mc, err := ctrl.client.MachineconfigurationV1().MachineConfigs().Get(context.TODO(), managedKey, metav1.GetOptions{})
+		isNotFound := errors.IsNotFound(err)
+		if err != nil && !isNotFound {
+			return fmt.Errorf("error checking for %s machine config: %v", managedKey, err)
+		}
+		// continue to the next MC if this already exists
+		if mc != nil && !isNotFound {
+			continue
+		}
+
+		tempIgnCfg := ctrlcommon.NewIgnConfig()
+		mc, err = mtmpl.MachineConfigFromIgnConfig(pool.Name, managedKey, tempIgnCfg)
+		if err != nil {
+			return fmt.Errorf("could not create capabilities MachineConfig from new Ignition config: %v", err)
+		}
+		rawCapsIgnition, err := json.Marshal(createNewIgnition(createDefaultCapsFile()))
+		if err != nil {
+			return fmt.Errorf("error marshalling capabilities config ignition: %v", err)
+		}
+		mc.Spec.Config.Raw = rawCapsIgnition
+		// Create the capabilities MC
+		if err := retry.RetryOnConflict(updateBackoff, func() error {
+			_, err = ctrl.client.MachineconfigurationV1().MachineConfigs().Create(context.TODO(), mc, metav1.CreateOptions{})
+			return err
+		}); err != nil {
+			return fmt.Errorf("could not create MachineConfig for capabilities: %v", err)
+		}
+		glog.Infof("Applied Capabilities MC %v on MachineConfigPool %v", managedKey, pool.Name)
+	}
+	return nil
+}
+
+// RunCapsBootstrap creates the capailities mc on bootstrap
+func RunCapsBootstrap(mcpPools []*mcfgv1.MachineConfigPool) ([]*mcfgv1.MachineConfig, error) {
```

Comment:
Do we need this function? I was under the impression that since you added a forced sync it will generate the capabilities MC that way.

This also looks like it just creates the machineconfig. What about the associated configmap?

Reply:
Yeah, we do need this. From what I understand, forcing the controller to sync is rendering an MC that has the capabilities stuff in it, but if we don't explicitly create this on bootstrap the cluster fails to start as the master nodes are trying to find a rendered MC that does not exist. The config map is created when the controller starts up, so there is no need to create it at bootstrap.